### ISSUE TITLE
Packages with same names in known repos and AUR

### ISSUE THREAD

**Issue by polygamma** [polygamma/aurman#23]
I do not know if this is only a theoretical problem, but it may be that a package in the AUR has the same name as a known repo package. What to do in this case? Prefer the repo package, prefer the AUR package, ask the user? Since I am new in the "AUR helper business" I don't have enough experience to decide that in a sensible way.

**Comment by rmarquis** [comment]
Assuming same name and samae version, this issue doesn't arise with the official repo and AUR.
It's different with user created repositories. In that case, it's best to assume the added repo takes priority above AUR packages: added repo are often (or always) added to avoid dealing with compiling from the AUR.

Explicit commands (such as `aurman -S aur/package` or `aurman -S reponame/package`) could be used to install one or the other.

**Comment by polygamma** [comment]
Thank you! I think it would be enough to make use of `aur/` then. It removes the need of knowing the name of the repo package. And as you said: Normally one wants to prefer the repo package, hence the user only has to write something else than the name, if he wants to explicitly install it from aur.

**Comment by polygamma** [comment]
https://github.com/polygamma/aurman/commit/835ad466b11efc7d68da444eea900aaf0bf6911d
what do you think?

**Comment by rmarquis** [comment]
You might want to add support for `reponame/package` too. This is especially useful when the user can choose between two similarly named packages available in binary repo. Checking the prefix exists but is not `aur/` is sufficient to the delegate the package to pacman; pacman will handle prefix errors itself too.

**Comment by polygamma** [comment]
Oh, that is possible, too? Meh... I am thinking about a clean solution then, instead of just adding that, too.

**Comment by AladW** [comment]
What happens when you have repo called `aur` in pacman.conf?

**Comment by polygamma** [comment]
just "/" for aur, maybe? that cannot conflict with names of repos

**Comment by polygamma** [comment]
How do I determine from which known repo a local package has been installed?

```
[jonny@jonny-arch-pc ~]$ pacman -Qi nano
Name            : nano
Version         : 2.9.3-1
Description     : Pico editor clone with enhancements
Architecture    : x86_64
URL             : http://www.nano-editor.org
Licenses        : GPL
Groups          : base
Provides        : None
Depends On      : ncurses  file  sh
Optional Deps   : None
Required By     : None
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 2.23 MiB
Packager        : Andreas Radke <andyrtr@archlinux.org>
Build Date      : Mon 29 Jan 2018 08:08:35 PM CET
Install Date    : Sun 04 Mar 2018 11:24:57 AM CET
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : Signature
```

```
[jonny@jonny-arch-pc ~]$ expac -Q '%n : %r' nano
nano : local
```

**Comment by AladW** [comment]
I don't think you can since pacman distinguishes between `/var/lib/pacman/local` and `/var/lib/pacman/sync`. I could be wrong, but it should make no difference whether you installed `nano` from `[extra]` or via `pacman -U`.

About the `/foo` suggestion, that makes me wonder if pacman accepts empty or null-byte repository names... it pretty much accepts anything else. In any case, it doesn't seem a very elegant solution to me. I guess an alternative is introduce an extra option like `--aur` or `--ignorerepo`.

**Comment by polygamma** [comment]
Let us assume the following:
I've listed 2 repos in my pacman.conf as follows:

[my_lovely_repo]
[i_love_you_more]

Now I want to install one package from my_lovely_repo which is also in i_love_you_more, hence I do this via `my_lovely_repo/package`. So far so good. Now I run `aurman -Syu` and there is a newer version of that package in i_love_you_more, hence aurman suggests to install this package, since aurman cannot know from which repo the package has been installed. But that would probably not be what the user wants. What to do? This really seems ugly and hacky as hell...

**Comment by AladW** [comment]
I can't test it right now; does `pacman -Syu` also suggest to install the newer version from the second repo?

**Comment by polygamma** [comment]
First: I'm an idiot. I got the order wrong, repos listed FIRST have higher priorities. So, what I meant is:

[i_love_you_more]
[my_lovely_repo]

to make the problem of my example clear.
You installed a package from a repo with LOWER prio and a new package becomes available in the repo with the HIGHER prio. I just added [testing] to my pacman.conf with HIGHEST prio, and surely pacman suggested to update a lot of packages when running -Syu.

So: If the user wants to keep a package from a repo with lower prio, how could one achieve that without dirty hacking, since it is not possible to find out from which repo the package has been installed?

**Comment by rmarquis** [comment]
@polygamma You can't and don't need to know which repo they've been installed from. Once a package is installed, it is a `local` package, which might or might not have a same package (name and version) in a sync repo.

Pacman will always respect the repo priorities (first repo in pacman.conf takes precedence), unless the version cannot be fullfilled in which case it will take the package in the next repo in line if it fulfils the version.

This also means issues could easily arise if the maintainers of these users repo don't do a good job at maintaining consistency with official repositories. This is exactly why I dislike using users repositories that duplicate official or AUR packages.

**Comment by polygamma** [comment]
Why should one even bother about allowing to install packages from repos that are "not the right one"?
I mean if one listed

[repo_high]
[repo_low]

both repos contain "package", hence pacman tries to install it from repo_high. Every -Syu would try to install it from repo_high, if the version of the package is higher in repo_high than from the installed package.

**Comment by polygamma** [comment]
I've thought about it for a while...
I do not think, that this is for the command line. If you want this, you want it not just once, you made the decision that you want to install package "x" from the aur or a special repository, from where it normally wouldn't be installed. Hence this is a thing for a config file.

So my idea is: There will a file, in which you can write names of packages and if you want to install them from aur or a special repo.

aurman will do the following:
all of those packages will be excluded from pacman -Syu by --ignore
when fetching the upstream packages for the dep solving, aurman will add the appropriate package.
when installing, and the package is listed as from aur, nothing special has to be done.
when installing, and the package is from a special repo, that will be appended to pacman with "repo/package".

does that seem sensible?

**Comment by rmarquis** [comment]
I do believe this is way more complex than it is needed. My opinion is: let users deal with it. That's the way pacman works, I don't think it should any different for AUR helpers.

If this is helping in any way, I never bothered supporting these users repository in pacaur, and I didn't get complaints that I can recall either.

**Comment by AladW** [comment]
>My opinion is: let users deal with it

Completely agreed. 

If anything you could support a `PACMAN` variable (or pass on `--config`) so that users can emulate the absence of these repos (if the package is installed from a repo, but said repo is not listed in pacman.conf, it is automatically foreign).

**Comment by polygamma** [comment]
To be honest I am not quite sure if we are talking about the same things right now.
What *exactly* do you mean by 

> My opinion is: let users deal with it. 

@rmarquis ?

Do you mean, that *nothing* should be changed, or add "aur/" and "repo/" flags but let the user deal with the problems which may arise, or something else?

**Comment by rmarquis** [comment]
The latter. Allowing to explicitly install a package from a repo is useful (pacman allows it with `-S reponame/package` too), but the fallout of dependency hell is none of your concerns.

On a sidenote, if you're using expac to load the packages in memory, you can use `expac -S -1` to force it to take the  first results only into account. It should be the highest priority package and should help in avoiding quite a few issues when enabling the `[testing]` repo or users repo.

**Comment by polygamma** [comment]
See: https://github.com/polygamma/aurman/tree/other_repos

I have decided to do what I wrote in: https://github.com/polygamma/aurman/issues/23#issuecomment-370943765

I understand your concerns about that being *too much* but I do not want to support things partially.
I see that installing from repos, from which things normally would not be installed, may be useful. Hence I want to support that, but as I said, I do not like supporting partially. Hence I implemented full support.

First things first: What the user yields via the config is being checked for correctness. It must not be, that packages are listed more than once, and also must all packages, which are being listed by the user, be available at where the user told us, that they are available. In case of an error, everything is being aborted and aurman quits with error message. So: I made sure that these things are correct.

Under this assumption the following is being done: All listed package names are being excluded from pacman -S in the beginning of the aurman routine by `--ignore`. These packages are not available for pacman in any way, if not given to pacman by aurman explicitly. Hence "normal pacman system updates" may not mess with those packages.

Next: When fetching the installed system packages, the things the user gave us, are taken into account. Hence the classification of the installed packages also yields what the user wants. We are nevertheless not able to tell from which source a package has been installed, hence we assume the source to be what the user gave us. For packages not listed by the user, nothing special will be done. 

When fetching the upstream system, the sources are taken into account, too. Hence aurman dep solving really works with the **correct** upstream packages.

Now for the sysupgrade: Since all packages listed by the user have been removed from the pacman -Syu, aurman handles upgrading for those packages. For aur packages this is surely already done, but those excluded repo packages are being handled by aurman for updating, too.

Installing: For aur packages nothing special has to be done, but for repo packages "repo_name/" is being put in front of the package name when giving the names to pacman, for explicitly listed packages by the user only.

**Comment by polygamma** [comment]
Full changes: https://github.com/polygamma/aurman/compare/f6908f174d2fdd044ff5b87fcc5d50b880ff2e03...6c02ba3db9f803f08fddf54cba390a910f7534d3

**Comment by AladW** [comment]
Comment on the right issue this time: 

I'm still not sure why this required such complex changes. You can by all means "exclude" packages from `pacman -Syu` by (temporary) removing the relevant repo/s from the pacman configuration.

**Comment by polygamma** [comment]
Well, I think that's a dirty hack, hence more partial support than anything else.
With the changes I made, it's nothing special anymore, just other packages in the upstream system than they would normally be. Okay, that one has to put "reponame/" in front of the name when installing with pacman is something special, but besides that, nothing. I also came up with some structure changes of the main.py while preparing this feature, hence aurman may now also do the upgrade for repo packages, so that it`s not "repo packages first, then aur packages" anymore, but "aur and repo packages together". but as i wrote in the readme: thats not recommended in general.
